hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/test/java/com/google/cloud/bigtable/hbase/replication/CloudBigtableReplicatorTest.java

```diff
@@ -437,4 +444,275 @@ public void testReplicateFailsToSubmitTask() throws IOException {
+    Cell cell1s =
+        new KeyValue(
+            getRowKey(1), CF1, DEFAULT_SOURCE_CBT_QUALIFIER.getBytes(), 0l, KeyValue.Type.Delete);
+
+    // Create replicator input WAL. Note that we don't have any expected replicator output.
+    BigtableWALEntry walEntry1 =
+        new BigtableWALEntry(TIMESTAMP, Arrays.asList(cell11, cell1s), TABLE_NAME_STRING);
+    Map<String, List<BigtableWALEntry>> walsToReplicate = new HashMap<>();
+    walsToReplicate.put(TABLE_NAME_STRING, Arrays.asList(walEntry1));
+
+    // Everything succeeds.
+    when(mockExecutorService.submit((Callable<Object>) any()))
+        .thenReturn(CompletableFuture.completedFuture(true));
+
+    // Trigger replication
+    assertTrue(replicator.replicate(walsToReplicate));
+
+    // called during constructor
+    verify(mockMetricExporter).incCounters(INCOMPATIBLE_MUTATION_METRIC_KEY, 0);
+    verify(mockMetricExporter).incCounters(DROPPED_INCOMPATIBLE_MUTATION_METRIC_KEY, 0);
+    verify(mockMetricExporter).incCounters(INCOMPATIBLE_MUTATION_DELETES_METRICS_KEY, 0);
+    verify(mockMetricExporter).incCounters(INCOMPATIBLE_MUTATION_TIMESTAMP_OVERFLOW_METRIC_KEY, 0);
+    verify(mockMetricExporter).incCounters(PUTS_IN_FUTURE_METRIC_KEY, 0);
+
+    // WAL log is empty because all entries were dropped. Refer metrics for test success.
+    // verify(mockExecutorService)
+    //     .submit(
+    //         new CloudBigtableReplicationTask(
+    //             TABLE_NAME_STRING, mockConnection, new HashMap<>()));
+    verify(mockMetricExporter).incCounters(SOURCE_CBT_DROPPED, 1);
+
+    Mockito.verifyNoMoreInteractions(mockMetricExporter, mockExecutorService);
+    Mockito.verifyNoInteractions(mockConnection);
+  }
+
+  /**
+   * Two-way replication combined test to check that replicator can drop the externally-replicated
+   * mutation but keep and replicate its own mutation.
+   *
+   * @throws IOException
+   */
+  @Test
+  public void testTwoWayReplicationDropsOneReplicatesOther() throws IOException {
+    // Create object to test
+    CloudBigtableReplicator replicator = new CloudBigtableReplicator();
+    replicator.start(sharedResources, incompatibleMutationAdapter, 2000, false);
+
+    // Enable two-way replication manually.
+    Configuration replicationConf = new Configuration(false);
+    replicationConf.setBoolean(ENABLE_TWO_WAY_REPLICATION_MODE_KEY, true);
+    replicator.maybeStartTwoWayReplication(replicationConf, mockMetricExporter);
+
+    // Row key 1 WAL comes from CBT and should be filtered out.
+    Cell cell11 = new KeyValue(getRowKey(1), CF1, null, TIMESTAMP, getValue(1));
+    // Special replicated mutation from CBT.
+    Cell cell1s =
+        new KeyValue(
+            getRowKey(1), CF1, DEFAULT_SOURCE_CBT_QUALIFIER.getBytes(), 0l, KeyValue.Type.Delete);
+
+    // Row key 2 WAL comes from HBase and should be replicated.
+    Cell cell21 = new KeyValue(getRowKey(2), CF1, null, TIMESTAMP, getValue(1));
+    Cell cell2s =
```

Comment:
better naming please. inputCell, expectedSourceTagCell?

Reply:
Renamed to be more informative